include/utils/AuxAccumulator.h

```diff
@@ -0,0 +1,87 @@
+//* This file is part of FENIX: Fusion Energy Integrated Multiphys-X,
+//* A multiphysics application for modeling plasma facing components
+//* https://github.com/idaholab/fenix
+//*
+//* FENIX is powered by the MOOSE Framework
+//* https://www.mooseframework.inl.gov
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "MooseTypes.h"
+
+#include "libmesh/fe.h"
+#include "libmesh/fe_map.h"
+
+class FEProblemBase;
+class AuxiliarySystem;
+template <typename T>
+class MooseVariableFE;
+
+namespace FENIX
+{
+/**
+ * Utility that accumulates values into an aux variable
+ */
+class AuxAccumulator
+{
+public:
+  AuxAccumulator(FEProblemBase & problem, const AuxVariableName & variable);
+  ~AuxAccumulator();
+
+  /**
+   * Accumulates a value into the aux field
+   *
+   * @param elem The element
+   * @param point The point at which to add the value (should be in \p elem)
+   * @param value The value at the point
+   */
+  void add(const Elem & elem, const Point & point, const Real & value);
+
+  /**
+   * Finalizes the accumulation; must be called once complete
+   */
+  void finalize();
+
+private:
+  /**
+   * Adds the cached values for the current element
+   *
+   * This enables not calling DofMap::dof_indices for each point that is added
+   * when successive points are in the same element
+   */
+  void addCachedValues();
+  /**
+   * Initializes the cached values for the given element
+   */
+  void initCachedValues(const Elem & elem, const std::size_t size);
+
+  /// The FE problem
+  FEProblemBase & _problem;
+
+  /// The mesh dimension
+  const unsigned int _dim;
+
+  /// The aux system
+  AuxiliarySystem & _aux;
+
+  /// The AuxVariable this AuxRayKernel contributes to
+  MooseVariableFE<Real> & _var;
```

Comment:
it looks like these are only set during the construction of the object shouldn't they be const as well? Or is there a reason they are not?

```suggestion
  const FEProblemBase & _problem;

  /// The mesh dimension
  const unsigned int _dim;

  /// The aux system
  const AuxiliarySystem & _aux;

  /// The AuxVariable this AuxRayKernel contributes to
  const MooseVariableFE<Real> & _var;
```

Reply:
I think you're forgetting what `const` means for a reference. A `const` reference means that you can only call `const` methods on that class. 

The only place where that's possible here is with `_var`, which probably can be const. On `_problem` we call non-const mesh accessors and on the aux system we call `solution()` which is not const because we're changing the solution.

Reply:
That makes sense thank you, I'm still learning all of the ins and outs of C++ so thank you for the explanation!

Reply:
Yep, sure thing! 

`_var` is now const in the above, so that one was good 👍🏼